lnwallet/channel_test.go

```diff
@@ -574,6 +578,234 @@ func TestSimpleAddSettleWorkflow(t *testing.T) {
+	}
+	defer cleanUp()
+
+	// Check that weight estimation of the commitment transaction without
+	// HTLCs is right.
+	checkSize(aliceChannel, 0)
+	checkSize(bobChannel, 0)
+
+	// Adding HTLCs and check that size stays in allowable estimation
+	// error window.
+	for i := 1; i <= 10; i++ {
+		htlc, _ := createHTLC(i)
+
+		if _, err := aliceChannel.AddHTLC(htlc); err != nil {
+			t.Fatalf("alice unable to add htlc: %v", err)
+		}
+		if _, err := bobChannel.ReceiveHTLC(htlc); err != nil {
+			t.Fatalf("bob unable to receive htlc: %v", err)
+		}
+
+		signNextCommitment(aliceChannel, bobChannel)
+		checkSize(aliceChannel, i)
+		checkSize(bobChannel, i)
+	}
+
+	// Settle HTLCs and check that estimation is counting cost of settle
+	// HTLCs properly.
+	for i := 10; i >= 1; i-- {
+		_, preimage := createHTLC(i)
+
+		settleIndex, err := bobChannel.SettleHTLC(preimage)
+		if err != nil {
+			t.Fatalf("bob unable to settle inbound htlc: %v", err)
+		}
+		err = aliceChannel.ReceiveHTLCSettle(preimage, settleIndex)
+		if err != nil {
+			t.Fatalf("alice unable to accept settle of outbound htlc: %v", err)
+		}
+
+		signNextCommitment(aliceChannel, bobChannel)
+		checkSize(aliceChannel, i-1)
+		checkSize(bobChannel, i-1)
+	}
+}
+
+// TestCheckHTLCNumberConstraint checks that we can't add HTLC or receive
+// HTLC if number of HTLCs exceed maximum available number, also this test
+// checks that if for some reason max number of HTLCs was exceeded and not
+// caught before, the creation of new commitment will not be possible because
+// of validation error.
+func TestCheckHTLCNumberConstraint(t *testing.T) {
+	createHTLC := func(i int) *lnwire.HTLCAddRequest {
+		preimage := bytes.Repeat([]byte{byte(i)}, 32)
+		paymentHash := fastsha256.Sum256(preimage)
+		return &lnwire.HTLCAddRequest{
+			RedemptionHashes: [][32]byte{paymentHash},
+			Amount:           lnwire.CreditsAmount(1e7),
+			Expiry:           uint32(5),
+		}
+	}
+
+	checkError := func(err error) error {
+		if err == nil {
+			return errors.New("Exceed max htlc count error was " +
+				"not received")
+		} else if err != ErrMaxHTLCNumber {
+			return errors.Errorf("Unexpected error occured: %v", err)
+		}
+
+		return nil
+	}
+
+	// Create a test channel which will be used for the duration of this
+	// unittest. The channel will be funded evenly with Alice having 5 BTC,
+	// and Bob having 5 BTC.
+	aliceChannel, bobChannel, cleanUp, err := createTestChannels(3)
+	if err != nil {
+		t.Fatalf("unable to create test channels: %v", err)
+	}
+	defer cleanUp()
+
+	// Add max available number of HTLCs.
+	for i := 0; i < MaxHTLCNumber; i++ {
+		htlc := createHTLC(i)
+		if _, err := aliceChannel.AddHTLC(htlc); err != nil {
+			t.Fatalf("alice unable to add htlc: %v", err)
+		}
+		if _, err := bobChannel.ReceiveHTLC(htlc); err != nil {
+			t.Fatalf("bob unable to receive htlc: %v", err)
+		}
+	}
+
+	// Next addition should cause HTLC max number validation error.
```

Comment:
Two additional checks should be added to ensure the `ReceiveHTLC` also triggers an error in this case.